### ISSUE TITLE
Implementation & static mut TS3_FUNCTIONS

### ISSUE THREAD

**Issue by 0xpr03** [ReSpeak/rust-ts3plugin#5]
Hello,
I'm a bit puzzled by the `static mut TS3_FUNCTIONS`.
For once there is a static reference to the Ts3Functions and all calls in TsApi are referring to it.
Not one is using any internal reference ( self.api... for example), so the whole struct seems kinda useless. Still you can't call any TS functions from any other thread, as you don't have a reference to this object as a user, though that's not entirely true, you can call the static logger function from any other thread, so it's inconsistent.
So my question is: What exactly is the reason for the const mut and still using some struct which has to be implemented, it feels wrong and makes it impossible to call functions from the outside (not ts event initiated).

**Comment by Flakebi** [comment]
Hi,
the initial meaning of the `TsApi` struct is to hold all (cached) data for the plugin and be the entry point for any function calls from the plugin to the TS api.
You are right that most the functions like `get_raw_api` and `get_…_path` there could be static.

The reason for the `static mut TS3_FUNCTIONS` is that the plugin gets this pointer once in the beginning, before the plugin is initialized, and it has to be stored somewhere. Also it cannot be stored in the `TsApi`, because the plugin-name is not yet available at this point, it gets set later.

One of the main points why I built this library is to have all the data available in an easy to use way – which is done through the caching. So far I did not need to access the cached data outside of the plugin events, so I just passed a reference to the `TsApi` to every function and all the mutex locking stays transparent for the user.
It should be easy to make the `Mutex<TsApi>` available to the plugin, as it is a static variable anyway, if that is what is missing for you?
